File: leet-code/maxProfitiii.js
//================递归=============================//
//================================================//
//================================================//
//================================================//

/**
 * @param {number[]} prices
 * @return {number}
 */
var maxProfit = function (prices) {
  if (prices.length === 0) {
    return 0
  }
  return dfs(prices, 0, 0, 0)
}

/**
 * @param {*} prices
 * @param {*} index 第几天
 * @param {*} status 状态 1-持有
 * @param {*} k 交易了几次
 */
var dfs = function (prices, index, status, k) {
  if (index === prices.length || k === 2) {
    return 0
  }

  // 定义三个变量分别记录 保持不动，买，卖
  let [a, b, c] = [0, 0, 0]
  a = dfs(prices, index + 1, status, k)

  if (status === 1) {
    // 如果持有一个股票，考虑卖出的情况
    b = dfs(prices, index + 1, 0, k + 1) + prices[index]
  } else {
    // 如果没有持有股票，考虑买入的情况
    c = dfs(prices, index + 1, 1, k) - prices[index]
  }
  // 返回三者的最大值
  return Math.max(a, b, c)
}

console.log(maxProfit([3, 3, 5, 0, 0, 3, 1, 4]))

//================记忆化搜索优化====================//
//================================================//
//================================================//
//================================================//

/**
 * @param {number[]} prices
 * @return {number}
 */
var maxProfit = function (prices) {
  const m = prices.length
  if (m === 0) {
    return 0
  }

  const profitMap = new Array(m).fill(0).map(item => {
    return new Array(2).fill(0).map(item => {
      return new Array(2).fill(0)
    })
  })
  return dfs(profitMap, prices, 0, 0, 0)
}

/**
 * @param {*} prices
 * @param {*} index 第几天
 * @param {*} status 状态 1-持有
 * @param {*} k 交易了几次
 */
var dfs = function (profitMap, prices, index, status, k) {
  if (index === prices.length || k === 2) {
    return 0
  }

  // 从缓存中找
  const cache = profitMap[index][status][k]
  if (cache > 0) {
    console.log('cache', cache)
    return cache
  }

  // 定义三个变量分别记录 保持不动，买，卖
  let [a, b, c] = [0, 0, 0]
  a = dfs(profitMap, prices, index + 1, status, k)

  if (status === 1) {
    // 如果持有一个股票，考虑卖出的情况
    b = dfs(profitMap, prices, index + 1, 0, k + 1) + prices[index]
  } else {
    // 如果没有持有股票，考虑买入的情况
    c = dfs(profitMap, prices, index + 1, 1, k) - prices[index]
  }
  // 返回三者的最大值
  const max = Math.max(a, b, c)
  // 存入缓存
  profitMap[index][status][k] = max
  return max
}

console.log(maxProfit([3, 3, 5, 0, 0, 3, 1, 4]))

File: leet-code/maxProfitiii.go
package main

// golang的递归+记忆化搜索也超时了。。。
func maxProfitii(prices []int) int {
	n := len(prices)
	profitMap := make([][][]int, n)
	for i := range profitMap {
		profitMap[i] = make([][]int, 2)
		for j := range profitMap[i] {
			profitMap[i][j] = make([]int, 2)
		}
	}

	var dfs func(profitMap [][][]int, prices []int, index, status, k int) int
	dfs = func(profitMap [][][]int, prices []int, index, status, k int) int {
		if index == len(prices) || k == 2 {
			return 0
		}

		cache := profitMap[index][status][k]
		if cache > 0 {
			return cache
		}

		a, b, c := 0, 0, 0
		a = dfs(profitMap, prices, index+1, status, k)
		if status == 1 {
			b = dfs(profitMap, prices, index+1, 0, k+1) + prices[index]
		} else {
			c = dfs(profitMap, prices, index+1, 1, k) - prices[index]
		}

		max := maxs(a, b, c)
		profitMap[index][status][k] = max
		return max
	}

	return dfs(profitMap, prices, 0, 0, 0)
}

// 动态规划
func maxProfitii2(prices []int) int {
	m := len(prices)
	// 一共有m天
	dp := make([][][]int, m)
	for i := 0; i < m; i++ {
		// 有两个状态，持有股票和未持有股票
		dp[i] = make([][]int, 2)
		for j := 0; j < 2; j++ {
			// 可以交易两次
			dp[i][j] = make([]int, 3)
		}
	}

	// 初始化数据
	for i := 0; i < 3; i++ {
		dp[0][1][i] = -prices[0]
	}

	for i := 1; i < m; i++ {
		for j := 0; j < 3; j++ {
			if j < 2 {
				dp[i][0][j] = max(dp[i-1][0][j], dp[i-1][1][j+1]+prices[i])
			} else {
				dp[i][0][j] = dp[i-1][0][j]
			}
			dp[i][1][j] = max(dp[i-1][1][j], dp[i-1][0][j]-prices[i])
		}
	}

	ans := 0
	for i := 0; i < 3; i++ {
		ans = max(ans, dp[m-1][0][i])
	}
	return ans
}
